crates/iceberg/src/spec/table_metadata.rs

```diff
@@ -101,7 +101,10 @@ pub const RESERVED_PROPERTIES: [&str; 9] = [
 /// Reference to [`TableMetadata`].
 pub type TableMetadataRef = Arc<TableMetadata>;
 
-#[derive(Debug, PartialEq, Deserialize, Eq, Clone)]
+#[derive(Debug, PartialEq, Deserialize, Eq, Clone, typed_builder::TypedBuilder)]
+#[builder(builder_method(name=declarative_builder))]
+#[builder(builder_type(name=TableMetadataDeclarativeBuilder, doc="Build a new [`TableMetadata`] in a declarative way. For imperative operations (e.g. `add_snapshot`) and creating new TableMetadata, use [`TableMetadataBuilder`] instead."))]
```

Comment:
I am not convinced "declarative" vs "imperative" is the defining distinction.
The `TableMetadataBuilder` is not "a builder pattern for TableMetadata". It's a facade to perform changes on the TableMetadata to be invoked by higher layer. For that reason, it includes _implicit changes_ that are supposed to be carried on when a metadata change happens, such as updating last_updated_ms when a new snapshot is added
https://github.com/apache/iceberg-rust/blob/3ea6f4791a2a4e15c4466cff7ccc478c08e2d3e3/crates/iceberg/src/spec/table_metadata_builder.rs#L394

The use-case described in the PR description is "the" standard builder pattern. To me, this is the defining difference. 

I suggest changing the doc string here and align the type names.

Reply:
I absolutely agree with this from a rust perspective. From an iceberg perspective other languages use the world `Builder` for this functionality unfortunately. I am uncertain if renaming the `TableMetadataBuilder` to a clearer name is worth the breaking change today.